sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala

```diff
@@ -184,9 +184,20 @@ object NestedColumnAliasing {
       plan: LogicalPlan,
       nestedFieldToAlias: Map[Expression, Alias],
       attrToAliases: AttributeMap[Seq[Alias]]): LogicalPlan = {
-    plan.withNewChildren(plan.children.map { plan =>
-      Project(plan.output.flatMap(a => attrToAliases.getOrElse(a, Seq(a))), plan)
-    }).transformExpressions {
+    val newChildPlan = plan match {
+      case g: Generate =>
+        g.withNewChildren(g.children.map { childPlan =>
+          val origOutput = childPlan.output
+          val fromAlias = childPlan.output.flatMap(a => attrToAliases.getOrElse(a, Nil))
```

Comment:
Hmm, we intend to replace some attributes with its nested fields if they are accessed on top of the plan. So we can prune unused fields later.

If we keep original outputs, I think pruning will not work actually.

Reply:
We may need to add a pruning test case to make sure pruning still works.

Reply:
I think, this change does not address the real issue. The real issue is, that Generate contains a list  unrequiredChildIndex of child output indices, that are not needed in the Generate output. This list has to be adjusted to fit the inserted Project node of NestedColumnAliasing. Here it just fits accidentally, because the original list is included completely at the beginning of the new Project node. But this may include unnecessary outputs, that ColumnPruning is trying to avoid. I have a different proposal, that adjust the list of indices to point to the new positions after attribute aliasing: https://github.com/apache/spark/pull/49061